### PR TITLE
OUT-782, OUT-783, OUT-558 Better drag & drop experience

### PR DESCRIPTION
**Pull request by aatbip**
### Changes

- [x] Refactors `DragDropHandler`
- [x] Adds a `CustomDragLayer` HOC to handle custom drag component for the list view task card
- [x] Adds a customized `getItemsStyle` to handle smooth dragging 
- [x] Adds `CardDragLayer` component as a custom drag layer for the list view and board view
- [x] Refactors design

### Testing criteria
- [x] [LOOM](https://www.loom.com/share/924a8625ade94d6b9e74e7dc5c16e643?sid=7e8d0ee2-5f5e-4c1d-9fa3-c39e3f454bb8)

### CODE REVIEW

**Comment by rrojan** [comment]
Please fix merge conflicts @aatbip

**Comment by aatbip** [comment]
> Please fix merge conflicts @aatbip

WARNING: There might be more merge conflicts coming forward after other PRs get merged. We can review PRs without merge conflict resolve. I will resolve conflicts before merging after the review is over. 
I am out for PM PTO today so won't be available to resolve conflicts if occur so please kindly do review even though conflicts occur. @rrojan

**Comment by rrojan** [comment]
> > Please fix merge conflicts @aatbip
> 
> WARNING: There might be more merge conflicts coming forward after other PRs get merged. I think we can review PRs without merge conflict resolve and conflicts can be resolved after review phase is over before continuing the merge. I am out for PM PTO today so won't be available to resolve conflicts if occurred so please kindly do review even though conflicts occur. @rrojan

Yes, but then we'd have to review it again after merge conflict resolution to make sure nothing is broken.

I'll be reviewing it as-is for now since you're away, but lets make sure to resolve conflicts before reviewing so we don't miss anything :bow:

**Review by rrojan — CHANGES_REQUESTED** [review]
@aatbip Is this PR incomplete, or is the card drop preview behavior not working?

Design:
image
image

Current behavior:
image

**Comment by aatbip** [comment]
> @aatbip The PR is incomplete - we don't have support for card preview on drop location.
> 
> Design: image
> 
> Current behavior: image
> 
> Same for list board


In the above Figma design, the destination droppable column is the left column since there is a border on it and the hover state is for the card being dragged from the start column.

Additionally, because we're sorting the cards based on their createdAt date, having a hover state on the droppable column might send a different signal, as hover states generally suggest the card will be dropped between other cards.

This PR was built around these considerations, and it also takes inspiration from Linear, where there's no hover state of task card, but they display a "board ordered by last updated" text in the board view. The list view, however, follows a design quite similar to ours.

cc @rrojan @makeitraina

**Comment by rrojan** [comment]
>In the above Figma design, the destination droppable column is the left column since there is a border on it and the hover state is for the card being dragged from the start column.

@aatbip The column that is currently being hovered over has the black border. It would be best to confirm this behavior before moving ahead :bow: 
cc @makeitraina

**Comment by aatbip** [comment]
> > In the above Figma design, the destination droppable column is the left column since there is a border on it and the hover state is for the card being dragged from the start column.
> 
> @aatbip The column that is currently being hovered over has the black border. It would be best to confirm this behavior before moving ahead 🙇 cc @makeitraina

The current implementation is correct. Please find the explanation here - https://linear.app/copilotplatforms/issue/OUT-558/better-drag-and-drop-experience#comment-3bb15dd3.

File: src/hoc/DragDropHandler.tsx [hunk 1]
```diff
@@ -1,58 +1,96 @@
...
+  moveCard,
+  onDropItem,
+  draggable = false,
+  droppable = false, // New prop for droppable
+}: Prop) => {
+  const { view } = useSelector(selectTaskBoard)
   const ref = useRef<HTMLDivElement | null>(null)
 
-  const [, drop] = useDrop({
+  const [{ isOver, canDrop }, drop] = useDrop({
```
Comment (priosshrsth) [inline src/hoc/DragDropHandler.tsx:37]: Can we do this instead? or something like this. Try to not use typecasting if possible.

const [{ isOver, canDrop }, drop] = useDrop<{ task: TaskResponse }, unknown, { isOver: boolean; canDrop: boolean }>({
    accept: accept,
    collect: (monitor: DropTargetMonitor) => ({
      isOver: monitor.isOver(),
      canDrop: monitor.canDrop(),
    }),
    drop: (item, monitor) => {
      if (onDropItem) {
        onDropItem({
          taskId: item.task.id,
          targetWorkflowStateId: id as string,
        })
      }
    },
    canDrop: () => droppable, // Only allow dropping if the component is droppable
  })

**Review by priosshrsth — CHANGES_REQUESTED** [review]

**Review by priosshrsth — CHANGES_REQUESTED** [review]

File: src/hoc/DragDropHandler.tsx (hunk 1 above)
Comment (aatbip) [reply src/hoc/DragDropHandler.tsx]: Fixed in the commit below

**Review by priosshrsth — APPROVED** [review]

**Pull request merged**